src/metabase/upload.clj

```diff
@@ -74,53 +75,38 @@
 ;;
 ;; </code></pre>
 
-(def ^:private type+parent-pairs
-  ;; listed in depth-first order
-  '([::boolean-or-int ::boolean]
-    [::boolean-or-int ::int]
-    [::auto-incrementing-int-pk ::int]
-    [::int ::float]
-    [::date ::datetime]
-    [::boolean ::varchar-255]
-    [::offset-datetime ::varchar-255]
-    [::datetime ::varchar-255]
-    [::float ::varchar-255]
-    [::varchar-255 ::text]))
-
-(defn ^:private column-type
-  "Returns the type of a column given the lowest common ancestor type of the values in the column."
-  [type]
-  (case type
-    ::boolean-or-int ::boolean
-    type))
-
-(def ^:private type->parents
-  (reduce
-   (fn [m [type parent]]
-     (update m type conj parent))
-   {}
-   type+parent-pairs))
+(def ^:private h
+  (-> (make-hierarchy)
+      (derive ::boolean-or-int ::boolean)
+      (derive ::boolean-or-int ::int)
+      (derive ::auto-incrementing-int-pk ::int)
+      (derive ::int ::float)
+      (derive ::date ::datetime)
+      (derive ::boolean ::varchar-255)
+      (derive ::offset-datetime ::varchar-255)
+      (derive ::datetime ::varchar-255)
+      (derive ::float ::varchar-255)
+      (derive ::varchar-255 ::text)))
+
+(def ^:private concrete-type?
+  "Nodes which can correspond to a concrete database column type."
+  (complement #{::boolean-or-int}))
```

Comment:
I find it strange to use insertion order to which type to fall back to. Maybe you have other reasons for it. Anyway, this is what I had in mind:

```clojure
(defn- specialise-ambiguous-type
  "For tags that correspond to more than one database type, specialise the tag so it corresponds to a single database type"
  {::boolean-or-int :boolean})
```

and then
```clojure
(defn ^:private column-type
  "The most specific column type corresponding to the given value type."
  [value-type]
  ;; If we know nothing about the value type, treat it as an arbitrary string.
  (cond (nil? value-type) ::text
        (specialise-ambiguous-type value-type) (specialise-ambiguous-type value-type)
        :else value-type))))
```

Reply:
Copied from the other PR: to me explicitly defining the parent type to use for types with multiple parents is preferable to making it depend on the insertion order. Why would you choose to couple these two things?

Another very tangible downside of this approach is you've had to implement an ordered hierarchy to achieve the same behaviour.

Reply:
I actually started out with exactly this approach, plus a test that we hadn't forgotten any entries in this map. To give some weight to your preference, this is how Clojure does things too - you specify these preferences per multi-method.

I'm quite firmly convinced this is the way to go though, and hope these reasons can persuade you as well:

1. For the linearization we use in the parser to be well defined, we need to specify an ordering for every parent, not just which is first. Perhaps we'll never have more than two parents for value-types, but an abstract `ordered-hierarchy` needs to handle this.
2. These preferences must be consistent with the ordering we use in the parser, and that seems error prone. If they're not consistent then you could get a different table from uploading one file or uploading it in two parts.
3. Having this map separate to the hierarchy means that we need an assert or unit test to ensure we haven't forgotten any entries.
4. Having this map separate to the derivations means that we can't maintain the ordering of all the internal nouns as an invariant.

The key thing I like about using order dependence like this is it establishes a 1-1 mapping between the derivation code and the graph visualization in our comment. We can just use that picture to reason about how our our type inference and relaxation works, and we can similarly see quickly how the picture maps to the code.

They key difference between Clojure's "a la carte" approach and ours, is that there isn't much value add for our hierarchy to be open - it makes sense in our case to define the full graph in one expression.

Reply:
While I really like using a `hierarchy` inspired data structure to power this graph, you've got me thinking that `make-hierarchy` and `derive` might not be the ideal interface for us to define the graph here - it's quite low level.

Perhaps you'd be less put off by a function that takes the full graph blueprint as arguments? It could be an ordered list of edges, which is a pretty shallow cosmetic change, or perhaps just a map. It would be *fairly* easy to turn a `parents` map into the corresponding derive statements (or even bypass them). I think a `children` map would be a bit more natural, at least it would map better to the visual depiction.

Taken to the extreme, we could even discard `derive` completely, and generate "closed" hierarchies that can't be extended. Even if you needed this extension you could still extract the parents map, modify that as a regular Clojure map, then pass that back into a hierarchy constructor.

Personally I find the "ordered list of edges" approach quite natural, and the original code before using a hierarchy already had an order dependency, in fact a more onerous one (all children before parents).

Reply:
We had an offline discussion, and here are the main points:

1. If we need to identify which types are abstract in any case. A map isn't much more code than a set, and provides nice documentation. We could document the same things in a unit test, but it's nice to have an explicit source of truth.
2. Using the ordering of the graph to define these projections was not adequately explained in the code, and is less obvious than an explicit map.
3. There is at least one important constraint we have between this projection and the graph structure. [^1]
4. There doesn't seem to be any constraint between the topological ordering and which parent we project an abstract type on to, for things to work out between importing, appending, etc with the current hierarchy.
5. I'm personally not convinced that we won't discover more important constraints as the hierarchy becomes more complex.
6. Testing the constraints is complex if the orderings are inconsistent, and guaranteed if they are.
7. Aesthetically its nice to normalize the shape of the graph, and read the implicit conversions right off its visualization.

Outcome:

1. We're bringing back the concrete map for now.
2. We'll test it satisfies the required constraints.

[^1]: There should be a route between every abstract type and its concrete type projection that consists only of abstract nodes.

Reply:
Thinking about abstract and concrete types at a higher level, I wonder whether we shouldn't late bind more on this projection. On a philosophical level, perhaps that are concrete for some drivers should be abstract for others, due to the differences in the granularity of their types. 

Today we would work around this by having driver methods generate the same definition for both nodes, but this feels hacky and error prone to me. It might actually be simpler for the driver not to implement that multimethod and fallback to nil, and for those nodes to be our implicit abstract nodes. We could continue walking the ancestors until we hit a type that the database does understand. This feels much simpler and robust to me, and would avoid us breaking existing drivers whenever we refine the hierarchy.

If we still wanted some drivers to simply break on some unsupported types rather than reading them as text, we could define that explicitly.

This is all very speculative around how more sophisticated the hierarchy could become in future.

Reply:
Yep. A concrete example of this is JSON columns. Not all our drivers support JSON/JSONB typed columns. One solution is to have the hierarchy and functions like `abstract->concrete` to depend on the driver.